### PR TITLE
 Set start of the week to Monday for root locale using SPI

### LINKED ISSUE #42588: Use ISO standard for week of week based year
In java.time week of weekbased year is depending on Locale. The locale we use is ROOT, and it does not define the start of the week. 
Therefore start of the week can be either Sunday or Monday.
ISO 8601 defines start of the week as Monday, and that standard was used in Joda for week of week based year.
This would fail because of that change. 
2019-21 (ok joda) vs 2019-22 (java)
```
 ZonedDateTime now = LocalDateTime.of(2019,5,26,1,32,8,328402)
                                         .atZone(ZoneOffset.UTC);
        DateFormatter jodaFormatter = Joda.forPattern("xxxx-ww").withLocale(Locale.ROOT).withZone(ZoneOffset.UTC);
        DateFormatter javaFormatter = DateFormatter.forPattern("8YYYY-ww").withLocale(Locale.ROOT).withZone(ZoneOffset.UTC);
        assertThat(jodaFormatter.format(now), equalTo(javaFormatter.format(now)));
```

### LINKED ISSUE #43275: Elasticsearch 7 reports a different key_as_string for dayOfWeek for date_histogram aggs than Elasticsearch 6 did
**Elasticsearch version** (`bin/elasticsearch --version`): 6.7.2 and 7.1.1

**Plugins installed**: []

**JVM version** (`java -version`): Java HotSpot(TM) 64-Bit Server VM/11.0.1/11.0.1+13-LTS

**OS version** (`uname -a` if on a Unix-like system): Windows 10

**Description of the problem including expected versus actual behavior**:

We are looking at upgrading from Elasticsearch 6.x to 7.x and one of our tests related to day-of-week in date-histogramm aggregations is failing. When investigating a bit more we found that there is actually a different response for the "key_as_string" for buckets of a date_histogramm aggregation which aggregates on "day of week" in Elasticsearch 7 compared to Elasticsearch 6. 

I could not find any "breaking change" mentioned for this, so I believe this may be an unintended side-effect of switching from Joda Time to Java DateTime API (https://www.elastic.co/guide/en/elasticsearch/reference/7.0/breaking-changes-7.0.html#breaking_70_java_time_changes)?

**Steps to reproduce**:

```
curl -X PUT "localhost:9200/twitter/_doc/1" -H 'Content-Type: application/json' -d'
{
    "user" : "kimchy",
    "post_date" : "2009-11-15T14:12:12",
    "message" : "trying out Elasticsearch"
}
'
```

```
curl -XPOST 'http://localhost:9200/twitter/_search' -d '{
 "size": 0,
 "aggregations": {
  "q2": {
   "date_histogram": {
    "field": "post_date",
    "format": "e",
    "interval": "day",
    "offset": 0
   }
  }
 }
}'
```

On Elasticsearch 6.7.2 this returns:

```
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 0,
    "hits": [

    ]
  },
  "aggregations": {
    "q2": {
      "buckets": [
        {
          "key_as_string": "7",
          "key": 1258243200000,
          "doc_count": 1
        }
      ]
    }
  }
}
```

however on Elasticsearch 7.1.1 is returns (note the difference in "key_as_string":

```
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": null,
    "hits": [

    ]
  },
  "aggregations": {
    "q2": {
      "buckets": [
        {
          "key_as_string": "1",
          "key": 1258243200000,
          "doc_count": 1
        }
      ]
    }
  }
}
```

**Provide logs (if relevant)**:

### PR DESCRIPTION
**Pull request by pgomulka**
Set start of the week to Monday for root locale (#43652)  …
Introducing a IsoLocal.ROOT constant which should be used instead of java.util.Locale.ROOT in ES when dealing with dates. IsoLocal.ROOT  customises start of the week to be Monday instead of Sunday.

closes #42588 an issue with investigation details
relates #41670 bug raised (this won't fix it on its own. joda.parseInto has to be reimplemented
closes #43275 an issue raised by community member

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
One question

File: distribution/src/config/jvm.options [hunk 1]
```diff
@@ -111,9 +111,11 @@ ${error.file}
 8:-XX:+UseGCLogFileRotation
 8:-XX:NumberOfGCLogFiles=32
 8:-XX:GCLogFileSize=64m
+# due to CalendarDataProvider being customized to ISO standard, SPI has to be first provider to try.
+8:-Djava.locale.providers=SPI,JRE
```
[the remark this answers is not in the document]
Comment (pgomulka) [reply distribution/src/config/jvm.options]: will do

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java [hunk 2]
```diff
@@ -39,6 +40,28 @@
 import static org.hamcrest.Matchers.is;
 
 public class JavaJodaTimeDuellingTests extends ESTestCase {
+    @Override
+    protected boolean enableWarningsCheck() {
+        return false;
```
[the remark this answers is not in the document]
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: will have a look to enable this again

**Comment by pgomulka** [comment]
the SPI mechanism works fine for jdks 9+. However it won't work for jdk8 in the same way. Before jdk9 only classes loaded with java extension mechanism could be used as SPI for packages `java.util.*`. The class we want to load with SPI is `java.util.spi.CalendarDataProvider`

So for jdk8 we would have to either place the jar with the SPI class in the `lib/ext` or made sure that we use `java.ext.dirs`

The JDK 9 release notes mentioning this change [Release Note: Locale sensitive SPI and Input Method SPI implementations are now loaded from application's classpath](https://bugs-stage.openjdk.java.net/browse/JDK-8167324)

The bug raised against jdk9 to fix this [Support java.util.spi.*, java.text.spi.*, java.awt.im.spi loaded from classpath](https://bugs-stage.openjdk.java.net/browse/JDK-8062588)

SPI loading in jdk8 https://github.com/bpupadhyaya/openjdk-8/blob/master/jdk/src/share/classes/sun/util/locale/provider/SPILocaleProviderAdapter.java#L80
SPI loading in jdk9  https://github.com/AdoptOpenJDK/openjdk-jdk9/blob/master/jdk/src/java.base/share/classes/sun/util/locale/provider/SPILocaleProviderAdapter.java#L81
Note different classloaders

**Comment by pgomulka** [comment]
closed in favour of https://github.com/elastic/elasticsearch/pull/48349

**Comment by pgomulka** [comment]
closed in favour of #48349

**Pull request closed without merging**